test/e2e/deployment.go

```diff
@@ -0,0 +1,511 @@
+	HTTPProxyCRD              *apiextensions_v1.CustomResourceDefinition
+	TLSCertDelegationCRD      *apiextensions_v1.CustomResourceDefinition
+	CertgenServiceAccount     *v1.ServiceAccount
+	ContourRoleBinding        *rbac_v1.RoleBinding
+	CertgenRole               *rbac_v1.Role
+	CertgenJob                *batch_v1.Job
+	ContourClusterRoleBinding *rbac_v1.ClusterRoleBinding
+	ContourClusterRole        *rbac_v1.ClusterRole
+	ContourService            *v1.Service
+	EnvoyService              *v1.Service
+	ContourDeployment         *apps_v1.Deployment
+	EnvoyDaemonSet            *apps_v1.DaemonSet
+
+	// Ratelimit deployment.
+	RateLimitDeployment       *apps_v1.Deployment
+	RateLimitService          *v1.Service
+	RateLimitExtensionService *contourv1alpha1.ExtensionService
+}
+
+// Unmarshals resources from rendered Contour manifest in order.
+// Note: This will need to be updated if any new resources are added to the
+// rendered deployment manifest.
+func (d *Deployment) UnmarshalResources() error {
+	_, thisFile, _, ok := runtime.Caller(0)
+	if !ok {
+		return errors.New("could not get filename")
+	}
+	renderedManifestPath := filepath.Join(filepath.Dir(thisFile), "..", "..", "examples", "render", "contour.yaml")
+	file, err := os.Open(renderedManifestPath)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+	decoder := apimachinery_util_yaml.NewYAMLToJSONDecoder(file)
+
+	// Discard empty document.
+	if err := decoder.Decode(new(struct{})); err != nil {
+		return err
+	}
+
+	d.Namespace = new(v1.Namespace)
+	d.ContourServiceAccount = new(v1.ServiceAccount)
+	d.EnvoyServiceAccount = new(v1.ServiceAccount)
+	d.ContourConfigMap = new(v1.ConfigMap)
+	d.ExtensionServiceCRD = new(apiextensions_v1.CustomResourceDefinition)
+	d.HTTPProxyCRD = new(apiextensions_v1.CustomResourceDefinition)
+	d.TLSCertDelegationCRD = new(apiextensions_v1.CustomResourceDefinition)
+	d.CertgenServiceAccount = new(v1.ServiceAccount)
+	d.ContourRoleBinding = new(rbac_v1.RoleBinding)
+	d.CertgenRole = new(rbac_v1.Role)
+	d.CertgenJob = new(batch_v1.Job)
+	d.ContourClusterRoleBinding = new(rbac_v1.ClusterRoleBinding)
+	d.ContourClusterRole = new(rbac_v1.ClusterRole)
+	d.ContourService = new(v1.Service)
+	d.EnvoyService = new(v1.Service)
+	d.ContourDeployment = new(apps_v1.Deployment)
+	d.EnvoyDaemonSet = new(apps_v1.DaemonSet)
+	objects := []interface{}{
+		d.Namespace,
+		d.ContourServiceAccount,
+		d.EnvoyServiceAccount,
+		d.ContourConfigMap,
+		d.ExtensionServiceCRD,
+		d.HTTPProxyCRD,
+		d.TLSCertDelegationCRD,
+		d.CertgenServiceAccount,
+		d.ContourRoleBinding,
+		d.CertgenRole,
+		d.CertgenJob,
+		d.ContourClusterRoleBinding,
+		d.ContourClusterRole,
+		d.ContourService,
+		d.EnvoyService,
+		d.ContourDeployment,
+		d.EnvoyDaemonSet,
+	}
+	for _, o := range objects {
+		if err := decoder.Decode(o); err != nil {
+			return err
+		}
+	}
+
+	rateLimitExamplePath := filepath.Join(filepath.Dir(thisFile), "..", "..", "examples", "ratelimit")
```

Comment:
this and below is an addition on top of the base in https://github.com/projectcontour/contour/pull/3793